### ISSUE TITLE
when rtc comes up after systemd, it may "advance" system clock back to before the build-time epoch

### ISSUE THREAD

**Issue by xnox** [systemd/systemd#17737]
**systemd version the issue has been seen with**
 > v245.x
 > TODO double check with master

**Used distribution**
 > Ubuntu

**Linux kernel version used** (`uname -a`)
  > a weird 5.9.9+

**CPU architecture issue was seen on**
 > aarch64

**Expected behaviour you didn't see**
 > Nov 19 2020 12:24:25 ubuntu systemd[1]: System time before build time, advancing clock.
 > Feb 03 1970 05:35:54 ubuntu kernel: rtc-pm8xxx 200f000.spmi:pm8916@0:rtc@6000: registered as rtc0
 > Feb 03 1970 05:35:54 ubuntu kernel: rtc-pm8xxx 200f000.spmi:pm8916@0:rtc@6000: setting system clock to 1970-02-03T05:35:54 UTC (2871354)
 > *readvance again* Nov 19 2020 12:24:25 ubuntu systemd[1]: RTC time before build time, advancing clock, again.

**Unexpected behaviour you saw**
 > Nov 19 2020 12:24:25 ubuntu systemd[1]: System time before build time, advancing clock.
 > Feb 03 1970 05:35:54 ubuntu kernel: rtc-pm8xxx 200f000.spmi:pm8916@0:rtc@6000: registered as rtc0
 > Feb 03 1970 05:35:54 ubuntu kernel: rtc-pm8xxx 200f000.spmi:pm8916@0:rtc@6000: setting system clock to 1970-02-03T05:35:54 UTC (2871354)

**Steps to reproduce the problem**
So it looks like the relevant RTC clock module is not built into the kernel, but is loaded at runtime from the initrd. It also looks like the battery is busted, or the time on the board is set to 1970 cause it was never booted. Thus I am observing that systemd starts in the initrd, advances clock to build-time epoch, then rtc comes up, and rewinds the time back to 1970. And then things fail in the initrd trying to do signature verification as some CA certs are not valid yet.

It would be nice if systemd would monitor clock change event, and readvance it to build-epoch.

Not sure if that is the task of timesyncd or systemd itself.... especially since this initrd does not establish network access.

**Is it a kernel bug?**
Well, it would be nice, if kernel would do sanity check, and if clock is at kernel-built-time-epoch (same as systemd epoch), or if it was advanced to systemd time epoch, that we should probably not go back to 1970. It would nice to do sanity check on rtc clock, and encode the build-time into the kernel. Such that we would ignore RTC clocks that talk about things before kernel built-time.

**Comment by keszybz** [comment]
I'd say that this is a bug in the rtc kernel drivers. The kernel knows that the system time has already been set, and it should not set it from rtc if it has already been set by userspace. Or alternatively, it should never move system time backwards based on rtc. Since by default time starts at the epoch, if the system time is bigger than the rtc time, it's almost certain that the rtc time is wrong. (Or neither can be trusted, in which case not doing anything is still better.)

**Comment by poettering** [comment]
I wasn't aware that loading RTC modules will adjust the system clock all kernel internal. Are you sure that that's what happens?

I assume the in-kernel system clock setting happens only for compiled in modules, before first userspace is spawned. It would really really suck if the kernel would adjust the system clock automatically even after userspace is already initialized and made some changes...

**Comment by xnox** [comment]
> I wasn't aware that loading RTC modules will adjust the system clock all kernel internal. Are you sure that that's what happens?
> 
> I assume the in-kernel system clock setting happens only for compiled in modules, before first userspace is spawned. It would really really suck if the kernel would adjust the system clock automatically even after userspace is already initialized and made some changes...

It used to be built in only, until this commit added support for runtime loadable rtc modules https://github.com/torvalds/linux/commit/f9b2a4d6a5f18e0aaf715206a056565c56889d9f

And indeed that breaks the assumption that systemd can push the clock to at least "build-time epoch" and that we would stay roughly in the right year, without NTP.

**Comment by xnox** [comment]
I'm pondering if it should check TK_CLOCK_WAS_SET before trying to do do_settimeofday64() upon loading the module.

**Comment by xnox** [comment]
I don't do kernels, will test build https://github.com/xnox/linux/commit/1ed80947460757df4c3090b42b9a47cfb3a3e821

**Comment by poettering** [comment]
smells like a compat breakage on the kernel side to just enable this logic suddenly...

why is the rtc module on this system compiled as module? does that make sense?

**Comment by xnox** [comment]
> smells like a compat breakage on the kernel side to just enable this logic suddenly...
> 
> why is the rtc module on this system compiled as module? does that make sense?

the original public commit is vague..... I can only guess that maybe they have multiple competing rtc drivers thus can't all be compiled in at the same time. Or like one wants to share the small kernel image and only load the desired/matching the platform rtc driver and have many platforms. In general, it is nice I guess to have everything as a module.

Re compat break - I am using systemd in the initrd hence noticed this and clearly have a dead battery in the rc. But majority of systems wouldn't notice. As things would come up in the right order, or the clock would be working one.

I actually was looking to submit a patch to the kernel to do, what systemd does with built-time epoch. Kernel too, should take the build time epoch and set the clock to at least that.

**Comment by xnox** [comment]
Actually this must be for the modular android - were upstream builds and ship kernel image and 3rd party manufacturers ship modules only to enable a particular device.

**Comment by alexandrebelloni** [comment]
I replied to the patch that was just sent to the kernel. I still disagree that the kernel should be setting the system time at all. My point of view is that systemd is better equipped to actually read the RTC at boot and set the system time if it decides that it is valid.

Using hctosys is basically only working properly on x86, because all the distributions have rtc0 as RTC_HCTOSYS_DEVICE and this is in most cases not the RTC that should be used.

In short, choosing which RTC is the correct one, reading it and setting system time should be done by userspace. Mandating hctosys was a wrong design decision. Notice that most distribution had an init script to read the RTC and set the system time before switching to systemd. My point being: let's fix that properly in systemd so we stop having those kind of issues popping up. Then we will be able to kill hctosys.

**Comment by alexandrebelloni** [comment]
To be clear, the proposed patch may seem fine but the kernel doesn't actually have the necessary information to take that decision, it doesn't know where the system time is coming from and whether the RTC is more correct.

Also, I can predict that the next patch coming in is "do not set the system time when NTP is enabled". Again, all of this could be avoided if done from userspace.

**Comment by poettering** [comment]
> In short, choosing which RTC is the correct one, reading it and setting system time should be done by userspace. Mandating hctosys was a wrong design decision. Notice that most distribution had an init script to read the RTC and set the system time before switching to systemd. 

The main point of which was to do the local-time-vs.-UTC jump. Which systemd still does today, just not in an init script. It's done in PID 1 super-early on. There's no functionality loss with systemd compared to the status quo ante.

> My point being: let's fix that properly in systemd so we stop having those kind of issues popping up. Then we will be able to kill hctosys.

I am not sure why you think systemd could do this better than the kernel. What do you think systemd can do the kernel can't anyway? Note that RTC is not something people ever want to configure explicitly, it's stuff that just *has* *to* *work* without any configuration. It sucks hard that we currently have to store the local-time-vs.-UTC jump in a file in /etc (which thus means it can only be applied after the initrd transition), but at least there we can say that you don't want that functionality anyway, unless you care for compat with Windows.

In general I am pretty sure on pretty much every single subsystem its the same: the kernel should just do a reasonably default policy on its own, so that things just work in 90% of the cases. For the rest userspace can override it. Building a subsystem that is "broken by default" until userspace fixes it up sucks.

**Comment by keszybz** [comment]
> I don't do kernels, will test build [xnox/linux@1ed8094](https://github.com/xnox/linux/commit/1ed80947460757df4c3090b42b9a47cfb3a3e821)

What were the results?

> Separately, systemd should be sensitive for dynamically loaded rtc module, and when such a device appears i guess udev should start something like systemd-hwtosys@%i.service and like check things like if RTC clock is more in the future than EPOCH and there is no ntp running yet, i guess pick RTC time over the current one.

Sounds reasonable.

**Comment by poettering** [comment]
An instantiated service for this sounds weird, after all even if there are many RTCs, we only have one system clock, hence it sounds pointless every looking at more than a single one of the.

**Comment by alexandrebelloni** [comment]
> > In short, choosing which RTC is the correct one, reading it and setting system time should be done by userspace. Mandating hctosys was a wrong design decision. Notice that most distribution had an init script to read the RTC and set the system time before switching to systemd.
> 
> The main point of which was to do the local-time-vs.-UTC jump. Which systemd still does today, just not in an init script. It's done in PID 1 super-early on. There's no functionality loss with systemd compared to the status quo ante.
> 
Losing functionality was not my point but we actually lost the possibility to boot with a kernel without RTC_HCTOSYS without losing timedated functionality.

> > My point being: let's fix that properly in systemd so we stop having those kind of issues popping up. Then we will be able to kill hctosys.
> 
> I am not sure why you think systemd could do this better than the kernel. What do you think systemd can do the kernel can't anyway?

Well, the kernel can't know where the system time is coming from (systemd build time, ext4 last mount timestamp ?) and whether the RTC is more correct or not, it doesn't know whether NTP will be enabled. It can't afford to read the RTC accurately. systemd knows or could do all of that with very minimal effort.

> Note that RTC is not something people ever want to configure explicitly, it's stuff that just _has_ _to_ _work_ without any configuration.

I disagree here, this is something that some people want to be able to configure and currently, they can't because systemd is mandating RTC_HCTOSYS and all the distributions are setting RTC_HCTOSYS_DEVICE to rtc0. I don't feel like having to recompile the kernel is easier for any user.

> It sucks hard that we currently have to store the local-time-vs.-UTC jump in a file in /etc (which thus means it can only be applied after the initrd transition), but at least there we can say that you don't want that functionality anyway, unless you care for compat with Windows.

Don't tell me you would like the kernel to handle that, there is no way the kernel can know local vs UTC.

> In general I am pretty sure on pretty much every single subsystem its the same: the kernel should just do a reasonably default policy on its own, so that things just work in 90% of the cases. For the rest userspace can override it. Building a subsystem that is "broken by default" until userspace fixes it up sucks.

Well, my point is that the RTC subsystem shouldn't care about system time at all. It only have less information than systemd to take the correct decision.
I think you believe the hctosys device is being selected carefully by the kernel but it doesn't actually have any more information than userspace to select the proper device and it is being hardcoded. 

I really think the way forward would be for systemd to set the system time from the RTC at boot when it thinks this is the correct thing to do and have distribution stop enabling RTC_HCTOSYS.

**Comment by webczat** [comment]
> > > In short, choosing which RTC is the correct one, reading it and setting system time should be done by userspace. Mandating hctosys was a wrong design decision. Notice that most distribution had an init script to read the RTC and set the system time before switching to systemd.
> > 
> > 
> > The main point of which was to do the local-time-vs.-UTC jump. Which systemd still does today, just not in an init script. It's done in PID 1 super-early on. There's no functionality loss with systemd compared to the status quo ante.
> 
> Losing functionality was not my point but we actually lost the possibility to boot with a kernel without RTC_HCTOSYS without losing timedated functionality.
> 
> > > My point being: let's fix that properly in systemd so we stop having those kind of issues popping up. Then we will be able to kill hctosys.
> > 
> > 
> > I am not sure why you think systemd could do this better than the kernel. What do you think systemd can do the kernel can't anyway?
> 
> Well, the kernel can't know where the system time is coming from (systemd build time, ext4 last mount timestamp ?) and whether the RTC is more correct or not, it doesn't know whether NTP will be enabled. It can't afford to read the RTC accurately. systemd knows or could do all of that with very minimal effort.
> 
> > Note that RTC is not something people ever want to configure explicitly, it's stuff that just _has_ _to_ _work_ without any configuration.
> 
> I disagree here, this is something that some people want to be able to configure and currently, they can't because systemd is mandating RTC_HCTOSYS and all the distributions are setting RTC_HCTOSYS_DEVICE to rtc0. I don't feel like having to recompile the kernel is easier for any user.
> 
> > It sucks hard that we currently have to store the local-time-vs.-UTC jump in a file in /etc (which thus means it can only be applied after the initrd transition), but at least there we can say that you don't want that functionality anyway, unless you care for compat with Windows.
> 
> Don't tell me you would like the kernel to handle that, there is no way the kernel can know local vs UTC.
> 
> > In general I am pretty sure on pretty much every single subsystem its the same: the kernel should just do a reasonably default policy on its own, so that things just work in 90% of the cases. For the rest userspace can override it. Building a subsystem that is "broken by default" until userspace fixes it up sucks.
> 
> Well, my point is that the RTC subsystem shouldn't care about system time at all. It only have less information than systemd to take the correct decision.
> I think you believe the hctosys device is being selected carefully by the kernel but it doesn't actually have any more information than userspace to select the proper device and it is being hardcoded.
> 
> I really think the way forward would be for systemd to set the system time from the RTC at boot when it thinks this is the correct thing to do and have distribution stop enabling RTC_HCTOSYS.

I do agree, although I think the correct point is initramfs when you have one.
You could always make some initramfs setting to be able to enable/disable rtc configuration and somehow identify the correct device and then it sets it before journal starts, or something like that.
I also think building the kernel is rather not something you would like to do if othervise you could just edit few config files, even if this editing would also require rebuilding dtb. Of course people able to do both are able to build the kernel, but requiring them to do this is not necessarily a good thing long term.

**Comment by xnox** [comment]
i somehow see it as a udev rule that fires upon rtc device appearing and doing things to time the right way. With local knoweledge of ntp, local-utc time, and epoch/quick-checks.

Cause modular rtc devices is now a thing. In addition to built in one, or instead of built in one, and it is now racy between kernel & userspace.

Cause we do usually have more information in userspace of what time is probably closer to reality.

**Comment by xnox** [comment]
> > I don't do kernels, will test build [xnox/linux@1ed8094](https://github.com/xnox/linux/commit/1ed80947460757df4c3090b42b9a47cfb3a3e821)
> 
> What were the results?
> 

I don't have any, as I lost access to the machine in question. I'll try to regain it.

**Comment by poettering** [comment]
> I do agree, although I think the correct point is initramfs when you have one.

I vehemently disagree. Just because there are some outliers where there are multiple rtcs, the typical case is that a system just has one, and that that's the one the system clock shall be set from. i.e. you can cover > 90% of the cases trivially easily just by letting the kernel do the adjustment itself.

Yes, there are some corner cases like the one this bug is about, but I think this could be relatively simply be fixed in-kernel: i.e. if the kernel would only jump the system clock forward and never backwards, then the issue wouldn't even apply.

Also, plenty kernel subsystems track how things are configured. For example the network subsystem tracks for the names of interfaces how they came to be: by the kernel auto-picked, by userspace picked before the device was created, or by userspace after it was created. If you want, you could track that in kernel too: remember how the clock was set: by userspace or by kernel space, and you can then make a smarter decision about whether you still want to jump the kernel.

> You could always make some initramfs setting to be able to enable/disable rtc configuration and somehow identify the correct device and then it sets it before journal starts, or something like that.

"identify the correct device"? you make it sound as having 55 RTCs was the common case. it's not though: the common case is the vast majority having 1 RTC, then some having 0 RTCs, and the > 1 case only is in some specialist cases I'd guess. And even then: I think we can reasonably assume that RTCs are generally OK as time source, and if they aren't and are broken, they shouldn't be exposed at all, but just be blocked out way before anyone wants to configure things.

> I also think building the kernel is rather not something you would like to do if otherwise you could just edit few config files, even if this editing would also require rebuilding dtb. Of course people able to do both are able to build the kernel, but requiring them to do this is not necessarily a good thing long term.

There's some discrepancy in the proposal here: some people here appear to assume that userspace would have better configuration sources/hardware databases available than the kernel, and thus should be in power. Others otoh appear to think that HCTOSYS should be done in userspace before even the main initrd starts to run. But in that phase there simply is *no* udev,, there' is not host /etc, nothing. We have about as little info as the kernel itself. So what's the point?

**Comment by poettering** [comment]
> i somehow see it as a udev rule that fires upon rtc device appearing and doing things to time the right way. With local knoweledge of ntp, local-utc time, and epoch/quick-checks.

local-utc is a compat feature for Windows. We should not bother with that on anything that wasn't a PC that might run Windows. And in particular we shouldn't add support for that for any new infrastructure, since it's a terrible kludge.

**Comment by webczat** [comment]
> > I do agree, although I think the correct point is initramfs when you have one.
> 
> I vehemently disagree. Just because there are some outliers where there are multiple rtcs, the typical case is that a system just has one, and that that's the one the system clock shall be set from. i.e. you can cover > 90% of the cases trivially easily just by letting the kernel do the adjustment itself.
> 
> Yes, there are some corner cases like the one this bug is about, but I think this could be relatively simply be fixed in-kernel: i.e. if the kernel would only jump the system clock forward and never backwards, then the issue wouldn't even apply.
> 
> Also, plenty kernel subsystems track how things are configured. For example the network subsystem tracks for the names of interfaces how they came to be: by the kernel auto-picked, by userspace picked before the device was created, or by userspace after it was created. If you want, you could track that in kernel too: remember how the clock was set: by userspace or by kernel space, and you can then make a smarter decision about whether you still want to jump the kernel.
> 
> > You could always make some initramfs setting to be able to enable/disable rtc configuration and somehow identify the correct device and then it sets it before journal starts, or something like that.
> 
> "identify the correct device"? you make it sound as having 55 RTCs was the common case. it's not though: the common case is the vast majority having 1 RTC, then some having 0 RTCs, and the > 1 case only is in some specialist cases I'd guess. And even then: I think we can reasonably assume that RTCs are generally OK as time source, and if they aren't and are broken, they shouldn't be exposed at all, but just be blocked out way before anyone wants to configure things.
By identify I mean, some default to rtc0 but in case you want to change it, you can... I don't think there is a way to auto identify the one that is seen as correct without user configuration if it's not rtc0.
> 
> > I also think building the kernel is rather not something you would like to do if otherwise you could just edit few config files, even if this editing would also require rebuilding dtb. Of course people able to do both are able to build the kernel, but requiring them to do this is not necessarily a good thing long term.
> 
> There's some discrepancy in the proposal here: some people here appear to assume that userspace would have better configuration sources/hardware databases available than the kernel, and thus should be in power. Others otoh appear to think that HCTOSYS should be done in userspace before even the main initrd starts to run. But in that phase there simply is _no_ udev,, there' is not host /etc, nothing. We have about as little info as the kernel itself. So what's the point?

I only mentioned it should be started before journald on initramfs when you have one to make log timestamps correct. Of course normally that would mean udev is running before journald, so may not be a good idea? hmm

**Comment by alexandrebelloni** [comment]
> > I do agree, although I think the correct point is initramfs when you have one.
> 
> I vehemently disagree. Just because there are some outliers where there are multiple rtcs, the typical case is that a system just has one, and that that's the one the system clock shall be set from. i.e. you can cover > 90% of the cases trivially easily just by letting the kernel do the adjustment itself.

This is not correct, most of the devices running linux have more than one RTC. This is billions of embedded devices and android phones versus very few PCs.

**Comment by poettering** [comment]
> Well, the kernel can't know where the system time is coming from (systemd build time, ext4 last mount timestamp ?) and whether the RTC is more correct or not, it doesn't know whether NTP will be enabled. It can't afford to read the RTC accurately. systemd knows or could do all of that with very minimal effort.

I think you assume userspace works quite differently from how it actually works. Linux/systemd boots in three phases basically:

1. initrd – we have no configuration source available really, i.e. not access to /etc (restricted udev)
2. early boot – we have /etc, but no /var, i.e. we have access to local/UTC boolean, but not to any timestamp touch files (no udev)
3. late boot – we start having access to timestamp touch files (full udev)
(and even later comes the network, NTP and such, or maybe not)

The only thing systemd does in the initrd (step 1) is bump the clock to the build-time epoch, we have no other source of info. It's a "forward bump" only

The only thing we do on in early boot (step 2) is fix up the UTC/localtime delta. nothing else

The only thing we do in late boot (step 3) is do the touch file handling, which is also strictly a "forward bump".

Now, if the kernel would do HCTOSYS on the first RTC to show up, also in "forward" bump style, we should totally ok.

I also don't see what HCTOSYS has to do with NTP. NTP comes very very late into the game, i.e. after network configuration is up, after maybe we managed to connect to some network. NTP syncs are thus very sporadic events, and should never be required for normal operation, except in very specific user chosen conditions. Systems should be atonomous and have roughly correct clocks whenever possible, i.e. work independently of NTP.

**Comment by poettering** [comment]
> This is not correct, most of the devices running linux have more than one RTC. This is billions of embedded devices and android phones versus very few PCs.

And why are they exposed to Linux? What is userspace supposed to do with that?

**Comment by poettering** [comment]
> I disagree here, this is something that some people want to be able to configure and currently, they can't because systemd is mandating RTC_HCTOSYS and all the distributions are setting RTC_HCTOSYS_DEVICE to rtc0. I don't feel like having to recompile the kernel is easier for any user.

BTW, why not make this a kernel cmdline option? hctosysdev=… on the kernel cmdline? I mean, if you want this configurable, then make it configurable, but userspace is such a sad place for that, since user configuration (i.e. /etc) with working udev is available only so late during boot. And it wouldn't be any smarter that such a kernel cmdline option really.

**Comment by poettering** [comment]
> > It sucks hard that we currently have to store the local-time-vs.-UTC jump in a file in /etc (which thus means it can only be applied after the initrd transition), but at least there we can say that you don't want that functionality anyway, unless you care for compat with Windows.
> 
> Don't tell me you would like the kernel to handle that, there is no way the kernel can know local vs UTC.

As you can see right above what you wrote, I think it's a compat kludge, and am not too much concerned.

That said, I think it would actually be great if we wouldn't have to take care of that from userspace either. Things is, on UEFI systems, the firmware's clock API actually does return TZ offset info. (GetTime()'s `EFI_TIME` structure's `TimeZone` field). If the kernel would just use that, this would be so much less of a hassle for us...

**Comment by webczat** [comment]
note you can provide configuration to initramfs, also note initramfs can read kernel cmdline and actually does it. So... userspace can still do it with configuration in kernel cmdline.

**Comment by poettering** [comment]
> note you can provide configuration to initramfs, also note initramfs can read kernel cmdline and actually does it. So... userspace can still do it with configuration in kernel cmdline.

and why would that be a good idea? parse stuff out from the kernel, and immediately and unconditionally pass it back again? how pointless! why involve userspace in such games of ping pong? Just shortcut the damn trivial thing in the kernel

**Comment by webczat** [comment]
Note that I assume that rtcs are kernel modules, and for non pcs, for devices like rpi/whatever, they are if you don't rebuild kernel.
And in case something is loaded with a delay like kernel modules, userspace seems more appropriate to handle time being set. Although I may be just a bit wary of all this :)

**Comment by xnox** [comment]
I'd like to go back to just a common case of a single working RTC. And optionally another non battery backed one on the SOC (as it sometimes happens on the embedded boards).

In the above scenario there is only the one RTC that should be used. And I personally experienced that the kernel module for this RTC was not in the initrd, and only on the root filesystem. It fired, unpredictably, during systemd boot when it was modprobed from the root filesystem during boot.

Imho my patch the I proposed before to kernel should be applied in this instance, if nothing else, but because of a race that loadable rtc modules introduced when previously with the same kernel config there was none.

Separately I will go and try to inspect our devices to ensure that the non-working RTC are not exposed, cause it is silly to use them.

**Comment by xnox** [comment]
Looking at systemd code, it does link the first rtc0 to become /dev/rtc, unless kernel has some other one configured for the RTC_HCTOSYS. I don't see any sort of userspace RTC_HCTOSYS functionality in systemd / timesyncd. I'm now further convinced that since now rtc modules are loaded from userspace a guard must be added to not advance the time into the past. As normally userspace uses reliable out-of-band information as to which sort of decade we are in, even when offline, thus going back in time is counter productive and potentially harmful.

Such a check will not protect from a buggy RTC declaring that it is year 3006, but so be it.

**Comment by mvo5** [comment]
Fwiw, I spend a bit of time debugging the same issue on a PI. It seems there are (broadly) two options to fix this:
1. apply a kernel change like @xnox suggested in https://github.com/xnox/linux/commit/1ed80947460757df4c3090b42b9a47cfb3a3e821 
2. add a udev rule that runs when /dev/rtc appears that runs `src/shared/clock-util.c:clock_apply_epoch()` again

The downside of (2) is that it's racy, i.e. there will be a observable clock jump in userspace for a very short time.

**Comment by alexandrebelloni** [comment]
You forgot:

3. Stop using RTC_HCTOSYS
